utils/fast-math/logs.hpp

```diff
@@ -19,50 +17,133 @@
 #include <cmath>
 #include <ports-of-call/portability.hpp>
 
+/*
+ * These functions are for use when moving to and from the gridding space
+ * for Spiner grids.
+ *
+ * The core idea here is to move onto a grid that's roughly
+ * logarithmically spaced. It doesn't actually matter if that grid is
+ * EXACTLY logarithmic. Just that it is approximately so.
+ * There are however some constraints that do matter:
+ *
+ * 1. The function we use to translate to "grid space" must be
+ * precise, meaning you get the same answer each time.
+ *
+ * 2. The function must be invertible, meaning you can go back to
+ * "linear space." Ideally it is also continuous.
+ *
+ * 3. The function and its inverse must be fast.
+ *
+ * To meet these constraints, we approximate a logarithm by using
+ * frexp and ldexp to split a real number into its mantissa + exponent
+ * in base 2. The mantissa is a real number on the interval [0.5, 1)
+ * and the exponent is an integer such that
+ *
+ * mantissa * 2^exponent = number
+ *
+ * To truly take the log, one needs to take the log of the mantissa.
+ * However, the first term in the logarithm's Taylor series is linear.
+ * Thus a linear approximation to log, for small numbers is valid.
+ * We use that linear approximation, which turns out to be:
+ *
+ * 2*(mantissa - 1)
+ *
+ * So that mantissa = 0.5 returns log_2(0.5) = -1 and mantissa (1)
+ * yields 0.
+ *
+ * Then the approximate log of the whole thing is the approximation of
+ * the log of the mantissa, plus the exponent:
+ *
+ * 2*(mantissa - 1) + exponent
+ *
+ * Which is a continuous, invertible function that approximates lg
+ * only relatively accurately. The absolute difference of this
+ * function from lg is about 0.1 at its largest. This translates to a
+ * relative difference of at most 25%.
+ *
+ * However, we don't mind this difference, as we're only interested in
+ * generating a grid spacing "close enough" to log spacing. As long as
+ * we can go into and out of "grid space" reliably and quickly, we're
+ * happy, and this function is an EXACT map into and out of that
+ * space.
+ */
+
 namespace singularity {
-namespace Math {
+namespace FastMath {
 
 PORTABLE_FORCEINLINE_FUNCTION
-double log10(const double x) {
+double fastlg(const double x) {
+  int n;
+#ifndef SINGULARITY_USE_SINGLE_LOGS
+  // double precision is default
+  const double y = frexp(x, &n);
+#else
+  // faster but less accurate
+  const float y = frexpf((float)x, &n);
+#endif // SINGULARITY_USE_SINGLE_LOGS
+  return 2 * (y - 1) + n;
+}
 
-  // const double ILOG10 = 1./std::log(10.0);
-  constexpr double ILOG10 = 0.43429448190325176;
+PORTABLE_FORCEINLINE_FUNCTION
+double fastpow2(const double x) {
```

Comment:
Should there be some check that `x` is positive here and in `fastlg`?

Reply:
I can add an assert maybe, but I'd rather leave that to higher-level in the codes. This particular piece of code is supposed to be performant.

Reply:
Done.

Reply:
I guess the concern is that it gives back a number that is nonsense but probably doesn't look super wrong. The issue is only in `fastlg` I think (I was wrong before about `fastpow2` I think, this should work fine for negative powers), so I guess the user could be assumed to know you shouldn't pass a negative number to a log function.

Reply:
> I guess the user could be assumed to know you shouldn't pass a negative number to a log function.

I think this is reasonable to assume, but also no harm in an assert, as they are pre-processed away in release builds. I've added an assert.